File: method4/op_reshape.cpp
#include "op_reshape.h"

#include <array>
#include <functional>
#include <numeric>
#include <stdexcept>

namespace {
template <int Rank>
int64_t numel(const std::array<int64_t, Rank>& shape) {
  int64_t out = 1;
  for (int i = 0; i < Rank; ++i) {
    out *= shape[static_cast<size_t>(i)];
  }
  return out;
}

template <int OutRank>
std::array<int64_t, OutRank> infer_size(int64_t total,
                                        const std::array<int64_t, OutRank>& shape) {
  std::array<int64_t, OutRank> inferred = shape;
  int64_t known = 1;
  int64_t unknown = -1;
  for (int i = 0; i < OutRank; ++i) {
    if (shape[static_cast<size_t>(i)] == -1) {
      if (unknown != -1) {
        throw std::runtime_error("Only one inferred dimension allowed");
      }
      unknown = static_cast<int64_t>(i);
    } else {
      if (shape[static_cast<size_t>(i)] <= 0) {
        throw std::runtime_error("Invalid shape");
      }
      known *= shape[static_cast<size_t>(i)];
    }
  }

  if (unknown != -1) {
    if (total % known != 0) {
      throw std::runtime_error("Inferred dimension mismatch");
    }
    inferred[static_cast<size_t>(unknown)] = total / known;
  }

  if (numel<OutRank>(inferred) != total) {
    throw std::runtime_error("Reshape numel mismatch");
  }
  return inferred;
}

template <int InRank, int OutRank>
std::array<int64_t, OutRank> compute_view_stride(
    const std::array<int64_t, InRank>& shape,
    const std::array<int64_t, InRank>& stride,
    const std::array<int64_t, OutRank>& new_shape) {
  if (numel<InRank>(shape) != numel<OutRank>(new_shape)) {
    throw std::runtime_error("Reshape numel mismatch");
  }
  std::array<int64_t, OutRank> view_stride{};
  int64_t view_dim = static_cast<int64_t>(OutRank) - 1;
  int64_t chunk_base_stride = stride[static_cast<size_t>(InRank - 1)];
  int64_t tensor_numel = 1;
  int64_t view_numel = 1;

  for (int64_t tensor_dim = static_cast<int64_t>(InRank); tensor_dim-- > 0;) {
    tensor_numel *= shape[static_cast<size_t>(tensor_dim)];
    bool at_chunk_end = tensor_dim == 0;
    if (!at_chunk_end) {
      const auto td = static_cast<size_t>(tensor_dim);
      const auto prev = static_cast<size_t>(tensor_dim - 1);
      auto expected = shape[td] * stride[td];
      if (shape[prev] != 1 && stride[prev] != expected) {
        at_chunk_end = true;
      }
    }

    if (at_chunk_end) {
      while (view_dim >= 0 &&
             (view_numel < tensor_numel ||
              new_shape[static_cast<size_t>(view_dim)] == 1)) {
        view_stride[static_cast<size_t>(view_dim)] = chunk_base_stride * view_numel;
        view_numel *= new_shape[static_cast<size_t>(view_dim)];
        --view_dim;
      }
      if (view_numel != tensor_numel) {
        throw std::runtime_error("View not compatible");
      }
      if (tensor_dim > 0) {
        chunk_base_stride = stride[static_cast<size_t>(tensor_dim - 1)];
        tensor_numel = 1;
        view_numel = 1;
      }
    }
  }

  while (view_dim >= 0) {
    if (new_shape[static_cast<size_t>(view_dim)] != 1) {
      throw std::runtime_error("View not compatible");
    }
    view_stride[static_cast<size_t>(view_dim)] = chunk_base_stride * view_numel;
    view_numel *= new_shape[static_cast<size_t>(view_dim)];
    --view_dim;
  }

  return view_stride;
}
}  // namespace

template <int InRank, int OutRank>
TensorView<OutRank> reshape(const TensorView<InRank>& in,
                            const std::array<int64_t, OutRank>& target_shape) {
  TensorView<OutRank> out;
  out.ptr = in.ptr;
  out.shape = infer_size<OutRank>(numel<InRank>(in.shape), target_shape);
  out.stride = compute_view_stride<InRank, OutRank>(in.shape, in.stride, out.shape);
  return out;
}

template TensorView<8> reshape<8, 8>(const TensorView<8>&,
                                     const std::array<int64_t, 8>&);

File: method3/op_reshape.py
import math
from typing import Sequence, Tuple


def _infer_size(numel: int, shape: Sequence[int]) -> Tuple[int, ...]:
    inferred = []
    unknown_dim = -1
    known_product = 1
    for idx, dim in enumerate(shape):
        if dim == -1:
            if unknown_dim != -1:
                raise ValueError("Only one dimension can be inferred.")
            unknown_dim = idx
            inferred.append(-1)
        else:
            if dim <= 0:
                raise ValueError("Shape dims must be positive or -1 for infer.")
            known_product *= dim
            inferred.append(dim)

    if unknown_dim != -1:
        if numel % known_product != 0:
            raise ValueError("Inferred dimension is not integral.")
        inferred[unknown_dim] = numel // known_product

    if math.prod(inferred) != numel:
        raise ValueError("Reshape numel mismatch.")
    return tuple(int(dim) for dim in inferred)


def _compute_view_stride(
    shape: Sequence[int],
    stride: Sequence[int],
    new_shape: Sequence[int],
) -> Tuple[int, ...]:
    if math.prod(shape) != math.prod(new_shape):
        raise ValueError("Reshape numel mismatch.")
    if not shape:
        return ()

    view_stride = [0] * len(new_shape)
    view_dim = len(new_shape) - 1
    chunk_base_stride = stride[-1]
    tensor_numel = 1
    view_numel = 1

    for tensor_dim in range(len(shape) - 1, -1, -1):
        tensor_numel *= shape[tensor_dim]
        at_chunk_end = tensor_dim == 0
        if not at_chunk_end:
            prev_stride = stride[tensor_dim - 1]
            expected = shape[tensor_dim] * stride[tensor_dim]
            if shape[tensor_dim - 1] != 1 and prev_stride != expected:
                at_chunk_end = True

        if at_chunk_end:
            while view_dim >= 0 and (
                view_numel < tensor_numel or new_shape[view_dim] == 1
            ):
                view_stride[view_dim] = chunk_base_stride * view_numel
                view_numel *= new_shape[view_dim]
                view_dim -= 1
            if view_numel != tensor_numel:
                raise ValueError("Reshape is not viewable without copy.")
            if tensor_dim > 0:
                chunk_base_stride = stride[tensor_dim - 1]
                tensor_numel = 1
                view_numel = 1

    while view_dim >= 0:
        if new_shape[view_dim] != 1:
            raise ValueError("Reshape is not viewable without copy.")
        view_stride[view_dim] = chunk_base_stride * view_numel
        view_numel *= new_shape[view_dim]
        view_dim -= 1

    return tuple(view_stride)


def reshape_with_checks(
    shape: Sequence[int],
    stride: Sequence[int],
    target_shape: Sequence[int],
) -> Tuple[Tuple[int, ...], Tuple[int, ...]]:
    """Compute inferred shape + view-compatible stride (or raise)."""
    numel = math.prod(shape)
    inferred = _infer_size(numel, target_shape)
    target_stride = _compute_view_stride(shape, stride, inferred)
    return inferred, target_stride
